app/helpers/recent_items_helper.rb

```diff
@@ -0,0 +1,89 @@
+# A random selection of recently modified works to display as thumbnails
+# on the front page of the site.
+# The same selection is shown to all users at the same time.
+# The selection is reshuffled few minutes,
+# and is cached for speed.
+module RecentItemsHelper
+  class RecentItems
+    def initialize()
+        @how_many_works_to_show = 6
+        @how_often_to_change = 60 * 10 # ten minutes
+        @how_many_works_in_bag = 15
+    end
+
+    def recent_items()
+      trigger_reshuffle_if_needed()
+      # First, put a bunch of eligible works into a bag.
+      works_to_pick_from = bag_of_recent_items()
+      # Now, pick a few of these out of the bag at random to show.
+      # Reshuffle the bag every now and then.
+      srand get_num()
+      works_to_pick_from.sort_by{rand}[0... @how_many_works_to_show]
+    end
+
+    private
+    def trigger_reshuffle_if_needed()
+        if time_to_reshuffle?()
+          # Fetch a new bag of recent works
+          # from SOLR and reshuffle the bag
+          # even if there haven't been any new works added.
+          set_bag(nil) # thus forcing a new call to SOLR.
+        end
+    end
+
+    # @@num is a slowly incrementing integer that changes at most every
+    # how_often_to_change minutes.
+    def time_to_reshuffle?()
+      new_arbitrary_number = Time.now.to_i / @how_often_to_change
+      if num_missing? || get_num() != new_arbitrary_number
+        set_num(new_arbitrary_number)
+        return true
+      end
+      return false
+    end
+
+    def bag_of_recent_items
+      conds =  {
+        'read_access_group_ssim'=>'public'
+      }
+      opts = {
+        :rows=> @how_many_works_in_bag,
+        :sort=> ["system_modified_dtsi desc"]
+      }
+      set_bag(GenericWork.search_with_conditions(conds, opts)) if bag_missing?
+      get_bag()
+    end
+
+    # getter/setter methods for our two class variables:
+    def bag_missing?()
+      var_missing?(:@@bag) || (get_bag() == nil)
+    end
+    def set_bag (value)
+      self.class.class_variable_set(:@@bag, value)
+    end
+    def get_bag()
+      self.class.class_variable_get(:@@bag)
+    end
+
+
+    def num_missing?()
+      var_missing?(:@@num) || (get_num() == nil)
+    end
+    def set_num(value)
+      self.class.class_variable_set(:@@num, value)
+    end
+    def get_num()
+      self.class.class_variable_get(:@@num)
+    end
+
+    def var_missing?(symbol)
+      begin
+        tmp = self.class.class_variable_get(symbol)
+        defined?(tmp) != "local-variable"
+      rescue
+        true
+      end
+    end
```

Comment:
You don't really need this. 

a) if you get to that part of the code`defined?(tmp)` is actually _always_ gonna be `local-variable`, there's no purpose for the check. 
b) You don't need to get class variables as `self.class.class_variable_get`, unless you actually need some indirection/reflection. You can just ask for `@@class_variable`. 
c) I _thought_ you could just ask for `@@class_variable` and if it didn't already exist it would just return nil. I was surprised to try it in irb and see that indeed (unlike @instance_vars) that isn't the case!  Which probably motivated all this code...

HOWEVER. I think the right way to deal with it is... just to initialize all the class variables you want at the top of the class. 

```
class RecentItems
  # not positive what the num variable stores, can you rename this more descriptively too? Is it actually "last_refresh_timestamp"? 
  @@num = nil  
  @@bag = nil
```

If you just declare them like that, ruby will never complain about you accessing them, and you won't need all that self.class.class_variable_get/rescue business. Or the `_missing?` methods. 

_Really_ I'm not sure if you need any cover methods at all at that point. Do you need:

```ruby
def bag
  @@bag
end
```

Or can your code just access `@@bag` directly?  I think probably directly is fine for this little self-contained thing, although an argument could be made for the cover methods. BUT in ruby, when we do getters/setters, the convention is simply `def num` instead of `def get_num`, and `def num=(v)` instead of `def set_num`. 

All of this stuff is so self-contained in this case that we really _could_ just deploy it, it's not going to "infect" anything else. So if you're really sick of working on this, we could just do that, or you could take another pass and simplify the code a bit (remove a bunch of those methods!).  Your call.

Reply:
Or _I_ could take a pass at simplifying it to show you how I'd do it, as a compare and contrast. Whatever you like!

Reply:
No, I definitely should do it! I'll refactor based on your suggestions, which are great Stay tuned - I'll need a break from this collections bug soon, and this will be fun and easy by comparison! :)
Thanks for the great feedback.